tools/pkg/engine_repo_tools/lib/engine_repo_tools.dart

```diff
@@ -0,0 +1,228 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// A minimal library for discovering and probing a local engine repository.
+///
+/// This library is intended to be used by tools that need to interact with a
+/// local engine repository, such as `clang_tidy` or `githooks`. For example,
+/// finding the `compile_commands.json` file for the most recently built output:
+///
+/// ```dart
+/// final Engine engine = Engine.findWithin();
+/// final Output? output = engine.latestOutput();
+/// if (output == null) {
+///   print('No output targets found.');
+/// } else {
+///   final io.File? compileCommandsJson = output.compileCommandsJson;
+///   if (compileCommandsJson == null) {
+///     print('No compile_commands.json file found.');
+///   } else {
+///     print('Found compile_commands.json file at ${compileCommandsJson.path}');
+///   }
+/// }
+/// ```
+library;
+
+import 'dart:io' as io;
+
+import 'package:path/path.dart' as p;
+
+/// Represents the `$ENGINE` directory (i.e. a checked-out Flutter engine).
+///
+/// If you have a path to the `$ENGINE/src` directory, use [Engine.fromSrcPath].
+///
+/// If you have a path to a directory within the `$ENGINE/src` directory, or
+/// want to use the current working directory, use [Engine.findWithin].
+final class Engine {
+  /// Creates an [Engine] from a path such as `/Users/.../flutter/engine/src`.
+  ///
+  /// ```dart
+  /// final Engine engine = Engine.findWithin('/Users/.../engine/src');
+  /// print(engine.srcDir.path); // /Users/.../engine/src
+  /// ```
+  ///
+  /// Throws a [InvalidEngineException] if the path is not a valid engine root.
+  factory Engine.fromSrcPath(String srcPath) {
```

Comment:
Our conversation from lunch reminded me that passing a `FileSystem` from `package:file` here and to `findWithin` will let this library and its clients test with a `MemoryFileSystem` rather than the real thing.

Reply:
Yeah that's fair.

Here is a question though, is that particularly useful? For example, where this would be used (`clang_tidy`, `githooks`) there isn't a real way to make them work on a virtual file system - which means we'd be adding the abstraction to `Engine` for ... engine to test itself?

Dunno, seems sus - I'm not against it but I'd prefer to add it when it provides value?

Reply:
Sure, we can add it later if we need it.